Algorithms/Divide-and-Conquer/heap.c

```diff
@@ -0,0 +1,84 @@
+#include <stdio.h>
+#include<malloc.h>
+/*Name - Jyoti Mishra (jstjyoti)
+Email-jyotimishra29091997@gmail.com
+Heap-Sort (Divide and Conquer) 
+INTRODUCTION :  
+a. Definition of a complete Binary Tree: A  Tree is said to be complete binary tree if all it's levels have maximum number of nodes(two for each parent) except possibly the last level and 
+all the nodes in the last level appear as per left as possible. 
+b. Heap: A Heap is a Complete Binary Tree which is implemented using an array(Linear Data Structure). 
+Max Heap: A heap is called max heap if all the nodes having a value greater than every children of the node. 
+This is used for Sorting in Ascending Order. 
+Min Heap: a heap is called min heap if all the nodes having smaller value than every children of the node*/
+
+/*This function(swap()) is for swapping two values having Time Complexity O(1) and space Complexity O(1) as only one extra pointer required to swap values*/
+void swap(int *x,int *y)
+{
+	int t=*x;
+	*x=*y;
+	*y=t;
+}
+/*This function(heap()) is responsible for creation a max heap everytime when it is called .
+Space Complexity of this is O(1) as only a few single variables are needed and they together represent a constant number of extra variables
+Time Complexity of above funtion is O(Logn) as the task is divided into two equal halves T(n) depends on Height as we are taking one of left and right subtrees to adjust the tree structure
+Height of balanced tree having n nodes=Logn
+*/
+void heap(int *arr, int n, int i)
+{
+    int largest = i; 
+    int l = 2*i + 1;  
+    int r = 2*i + 2;  
+ 	if (l < n && *(arr+l) >*(arr+largest))
+        largest = l;
+ 	if (r < n && *(arr+r) > *(arr+largest))
+        largest = r;
+   if (largest != i)
+    {
+        swap((arr+i), (arr+largest));
+        heap(arr, n, largest);
+    }
+}
+/*This Function(heapSort()) is for Calling heap() functtion and causing sort, each time by removing last element after first and last elements are swapped.
+Space complexity of this is O(1) as each time only some constant number of Variables are needed.
+Time Coomplexity is O(nLogn) as it is calling heap() function(discussed complexity above the function) from an i loop having iteration n times(below i loop)
+first i loop has iterartions in order of n times so it is having complexity about O(nLogn)

```

Comment:
This loop which creates max-heap in lines 53-54 has time complexity less than O(n logn). Although this is sufficient to say that the complexity of heapSort is O(n log n) since the second for loop will take O(n logn) time. But, can you analyze further and find the precise complexity for just the loop?

Reply:
for the above one?

Reply:
For the first loop in the **heapSort** function which creates a max-heap.

Reply:
yes, I have tried to analyse operations and  explain.See once

Reply:
It is `O(n logn)` but a tighter bound is possible. In reality, this loop takes `O(n)` time if you analyze it carefully, since the **heap** function that you have defined doesn't always take **O(log n)** time but takes **O(h)** time where **h** is the height of tree at the node where you call it. (this node is the parameter _i_ in your call to heap function).

So, if you do the mathematical analysis, you end up at `O(N)` which is actually I expected you must know. Anyways, hope you learn it now.